File: bayesmendel-comparator/python/tests/test_comparison_builder.py
import pytest

from bayesmendel_comparator.comparison_builder import (
    ResponseSchemaError,
    build_comparison,
)
from tests.fixtures_loader import load_all_fixtures


def test_one_row_per_model_in_registry_order() -> None:
    table = build_comparison(load_all_fixtures())

    assert len(table.rows) == 3
    assert [row["Model"] for row in table.rows] == ["BRCAPRO", "MMRpro", "PancPRO"]


def test_columns_include_fixed_and_lifetime() -> None:
    table = build_comparison(load_all_fixtures())

    assert table.columns[0] == "Model"
    assert table.columns[1] == "Counselee"
    assert table.columns[2] == "Any carrier"
    assert table.columns[-1] == "Lifetime risk @max-age"


def test_gene_columns_unioned_in_first_seen_order() -> None:
    table = build_comparison(load_all_fixtures())

    gene_cols = [
        col
        for col in table.columns
        if col not in {"Model", "Counselee", "Any carrier", "Lifetime risk @max-age"}
    ]
    assert gene_cols == [
        "Pr(BRCA1 mutation)",
        "Pr(BRCA2 mutation)",
        "Pr(Both genes mutated)",
        "Pr(MLH1 mutation)",
        "Pr(MSH2 mutation)",
        "Pr(MSH6)",
    ]


def test_brcapro_row_has_brca1_and_no_mlh1() -> None:
    table = build_comparison(load_all_fixtures())

    brcapro = table.rows[0]
    assert brcapro["Pr(BRCA1 mutation)"] == 0.4239
    assert brcapro["Pr(MLH1 mutation)"] is None


def test_lifetime_summary_uses_oldest_age() -> None:
    table = build_comparison(load_all_fixtures())

    assert (
        table.rows[0]["Lifetime risk @max-age"]
        == "Breast Ca Risk 38.48%; Ovarian Ca Risk 28.46%"
    )


def test_missing_carrier_probabilities_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        build_comparison({"BRCAPRO": {"counselee_name": "X"}})


def test_non_numeric_carrier_probability_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        build_comparison(
            {
                "BRCAPRO": {
                    "counselee_name": "X",
                    "carrier_probabilities": {"Pr(BRCA1 mutation)": "oops"},
                    "future_risks": [],
                }
            }
        )


def test_empty_payload_mapping_raises() -> None:
    with pytest.raises(ResponseSchemaError):
        build_comparison({})

File: bayesmendel-comparator/r/tests/testthat/test-comparison_builder.R
test_that("produces one row per model in registry order", {
  comparison <- build_comparison(load_all_fixtures())

  expect_length(comparison$rows, 3L)
  expect_equal(
    vapply(comparison$rows, function(row) row$Model, character(1L)),
    c("BRCAPRO", "MMRpro", "PancPRO")
  )
})

test_that("columns include Model, Counselee, Any carrier and a lifetime column", {
  comparison <- build_comparison(load_all_fixtures())

  expect_equal(comparison$columns[[1L]], "Model")
  expect_equal(comparison$columns[[2L]], "Counselee")
  expect_equal(comparison$columns[[3L]], "Any carrier")
  expect_equal(tail(comparison$columns, 1L), "Lifetime risk @max-age")
})

test_that("unions gene columns across all three models in first-seen order", {
  comparison <- build_comparison(load_all_fixtures())

  gene_cols <- setdiff(
    comparison$columns,
    c("Model", "Counselee", "Any carrier", "Lifetime risk @max-age")
  )
  expect_equal(
    gene_cols,
    c("Pr(BRCA1 mutation)", "Pr(BRCA2 mutation)", "Pr(Both genes mutated)",
      "Pr(MLH1 mutation)", "Pr(MSH2 mutation)", "Pr(MSH6)")
  )
})

test_that("BRCAPRO row carries BRCA1 number and leaves MLH1 cell empty", {
  comparison <- build_comparison(load_all_fixtures())
  brcapro_row <- comparison$rows[[1L]]

  expect_equal(brcapro_row[["Pr(BRCA1 mutation)"]], 0.4239)
  expect_null(brcapro_row[["Pr(MLH1 mutation)"]])
})

test_that("lifetime summary picks the oldest age and formats percentages", {
  comparison <- build_comparison(load_all_fixtures())
  brcapro_row <- comparison$rows[[1L]]

  expect_equal(
    brcapro_row[["Lifetime risk @max-age"]],
    "Breast Ca Risk 38.48%; Ovarian Ca Risk 28.46%"
  )
})

test_that("missing carrier_probabilities raises response_schema_error", {
  expect_error(
    build_comparison(list(BRCAPRO = list(counselee_name = "X"))),
    class = "response_schema_error"
  )
})

test_that("non-numeric carrier probability raises response_schema_error", {
  bad <- list(BRCAPRO = list(
    counselee_name = "X",
    carrier_probabilities = list(`Pr(BRCA1 mutation)` = "oops"),
    future_risks = list()
  ))
  expect_error(build_comparison(bad), class = "response_schema_error")
})

test_that("empty payload list raises response_schema_error", {
  expect_error(build_comparison(list()), class = "response_schema_error")
})
